### ISSUE TITLE
plthook_exit error on qemu

### ISSUE THREAD

**Issue by gurugio** [namhyung/uftrace#130]
Hi,

I built my qemu with -pg option and checked it worked well.
I ran following command to debug my qemu process with uftrace.
```
sudo /home/gkim/uftrace/uftrace record /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64 \
-M pc-1.0 -enable-kvm -m 2048 -smp 2 -name pb-pserver \
-uuid e4b93f3f-e301-3a0b-1829-582cf34a6810 \
-monitor telnet:0.0.0.0:9400,server,nowait -rtc base=utc \
-drive file=/mnt/tmp/debian8.2-pserver.img,if=none,id=drive-virtio-disk1,format=raw,cache=none \
-device virtio-blk-pci,bus=pci.0,addr=0x6,drive=drive-virtio-disk1,id=virtio-disk1 \
-vnc 0.0.0.0:0
```

Following is the result of uftrace.

```
mcount: /home/gkim/uftrace/libmcount/plthook.c:688:plthook_exit
 ERROR: <18> invalid dynsym idx: 65535
```

Could you inform me what I did wrong?

**Comment by namhyung** [comment]
Hello,

The message shows that it hit an unmatched function entry and exit - precisely, entry was non-PLT function but it exited from a PLT function.  This can come from a non-local jump (setjmp/longjmp), user-level thread, co-routine or some other (asm?) magic.  This is because uftrace needs to hook return address to trace function exit (for `-pg` compiled binaries).  I need to look at which function causes the problem.  Could you please attach a log when you run `uftrace record -vv`?

To work-around the problem, you can try either:
1. recompile qemu with `-finstrument-functions`
2. disable tracing PLT functions by running uftrace with `--no-libcall`

**Comment by gurugio** [comment]
Unfortunately I couldn't build qemu with -finstrument-functions flag.

Following is the result of "uftrace record -vv": qemu built without -finstrument-functions

```
gkim@ib1:~/pb-qemu-kvm$ cat uftrace_go.sh 
#!/bin/bash
sudo /home/gkim/uftrace/uftrace record -vv /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64 \
-M pc-1.0 -enable-kvm -m 2048 -smp 2 -name pb-pserver \
-uuid e4b93f3f-e301-3a0b-1829-582cf34a6810 \
-monitor telnet:0.0.0.0:9400,server,nowait -rtc base=utc \
-drive file=/mnt/tmp/debian8.2-pserver.img,if=none,id=drive-virtio-disk1,format=raw,cache=none \
-device virtio-blk-pci,bus=pci.0,addr=0x6,drive=drive-virtio-disk1,id=virtio-disk1 \
-vnc 0.0.0.0:0
gkim@ib1:~/pb-qemu-kvm$ bash uftrace_go.sh 
[sudo] password for gkim: 
uftrace: removing uftrace.data.old directory
symbol: check trace functions in /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64
uftrace: creating 2 thread(s) for recording
uftrace: using libmcount.so library for tracing
uftrace: start writer thread 0
uftrace: start writer thread 1
mcount: initializing mcount library
symbol: loading dynamic symbols from /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64
mcount: opening executable image: /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64
mcount: found GOT at 0xa10f90 (PLT resolver: 0x7fd330a2c2b0)
mcount: overwrite [401] mcount: 0x7fd33080e420
mcount: mcount setup done
mcount: preparing shmem buffers
uftrace: MSG SESSION: 8546: /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64 (7b1f89dbcf174420)
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG TASK_START : 8546/8546
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: preparing shmem buffers
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8549-000
uftrace: MSG TASK_START : 8546/8549
mcount: preparing shmem buffers
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8550-000
uftrace: MSG TASK_START : 8546/8550
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: new buffer: [1] /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-001
mcount: new buffer: [0] /uftrace-7b1f89dbcf174420-8546-000
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-001
uftrace: MSG START: /uftrace-7b1f89dbcf174420-8546-000
mcount: setup jmpbuf rstack at 7ffc4f340cb0 (17 entries)
mcount: /home/gkim/uftrace/libmcount/plthook.c:688:plthook_exit
 ERROR: <18> invalid dynsym idx: 65535
mcount: shmem_finish: tid: 8546 seqnum = 44 curr = 0, nr_buf = 2 max_buf = 2
uftrace: MSG  END : /uftrace-7b1f89dbcf174420-8546-000
mcount: releasing all shmem buffers for task 8546
uftrace: MSG TASK_END : 8546/8546
uftrace: all process/thread exited
uftrace: child terminated with exit code: 1
uftrace: stop writer thread 1
uftrace: stop writer thread 0
uftrace: reading uftrace.data/task.txt file
session: new task: tid = 8546
session: new task: tid = 8549
session: new task: tid = 8550
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8549-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8550-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-001
uftrace: flushing /uftrace-7b1f89dbcf174420-8546-000
uftrace: unlink for session: 7b1f89dbcf174420
symbol: loading symbols from /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64
symbol: loading dynamic symbols from /home/gkim/pb-qemu-kvm/x86_64-softmmu/qemu-system-x86_64
symbol: saving symbols to uftrace.data/qemu-system-x86_64.sym
uftrace: chown uftrace.data directory to (1001:1001)
symbol: unload symbol tables
```

**Comment by namhyung** [comment]
Thanks for the report.  The following message showed that it was due to the setjmp/longjmp.

> mcount: setup jmpbuf rstack at 7ffc4f340cb0 (17 entries)
mcount: /home/gkim/uftrace/libmcount/plthook.c:688:plthook_exit
 ERROR: <18> invalid dynsym idx: 65535

Actually I added some code to detect it and recover current rstack from jmpbuf rstack.  But it seems not to work properly.  I will investigate it more..

**Comment by namhyung** [comment]
It seems qemu uses setjmp/longjmp extensively for TCG and coroutines.  Also I found that qemu is built with PIE + bind-now so it's related to #119 somehow.

**Comment by namhyung** [comment]
I took a look at the qemu's coroutine code.  It uses ucontext + setjmp/longjmp to implement coroutines and there's a problem in uftrace handling nested PLT calls.  This was because `swapcontext()` never returns and it prevents other PLT functions to be hooked by uftrace.  And setjmp/longjmp need special care for uftrace but it could not do that.

I pushed the fix to the [check/longjmp-fix](https://github.com/namhyung/uftrace/compare/check/longjmp-fix) branch.  Could you please test it?

**Comment by namhyung** [comment]
Merged: 83380b8

@gurugio it seems to work for me so I'd like to close this issue.  If you could still find a problem, please re-open this.  Thanks,

**Comment by gurugio** [comment]
OK, I'll try it again and inform you if I have problem.
Thanks ;-)
